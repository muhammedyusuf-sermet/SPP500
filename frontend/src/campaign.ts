export interface ICampaignState {
	Id?: string,
	Name: string,
	Description: string;
	// Monsters: Monster.IMonsterState[];
}
