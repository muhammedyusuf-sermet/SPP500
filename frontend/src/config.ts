export const API_URL = "https://http://3.18.65.138:3000";