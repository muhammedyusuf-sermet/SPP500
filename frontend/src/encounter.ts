import { IMonsterData } from "./monster";

export interface IEncounterData {
	Id?: string,
	Name: string,
	Description: string;
	Monsters: IMonsterData[];
}
