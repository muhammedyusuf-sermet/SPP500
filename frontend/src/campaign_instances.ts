import * as Campaign from "./campaign";

// Dummy array of monsters
import EncounterInstances from './encounter_instances';

let encounterExample: Campaign.ICampaignState;
encounterExample = {
	Id: 0,
	Name: "Campaign",
	Summary: "Bloody hell...",
	Notes: "You gonna die!",
	Encounters: EncounterInstances['EncounterInstances'],
};

let CampaignInstances = new Array();

// Generate a dummy array of encounters with different IDs
for (var i = 0; i < 12; i++) {
	let encounter = {} as Campaign.ICampaignState;
	Object.assign(encounter, encounterExample);
	encounter['Id'] = i;
	CampaignInstances.push(encounter);
}

export default {
    CampaignInstances,
}