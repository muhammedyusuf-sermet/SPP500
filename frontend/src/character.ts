import { ICampaignData } from "./campaign";

export enum CharacterRace {
	Dragonborn = "Dragonborn",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Gnome = "Gnome",
	HalfElf = "HalfElf",
	Halfling = "Halfling",
	HalfOrc = "HalfOrc",
	Human = "Human",
	Tiefling = "Tiefling"
}

export enum CharacterClass {
	Barbarian = "Barbarian",
	Bard = "Bard",
	Cleric = "Cleric",
	Druid = "Druid",
	Fighter = "Fighter",
	Monk = "Monk",
	Paladin = "Paladin",
	Ranger = "Ranger",
	Rogue = "Rogue",
	Sorcerer = "Sorcerer",
	Warlock = "Warlock",
	Wizard = "Wizard"
}

export interface ICharacterData {
	Id?: number;
	Name: string;
	Level?: number;
	Race?: CharacterRace;
	Class?: CharacterClass;
	MaxHealth?: number;
	ArmorClass?: number;
	Notes?: string;

	Campaigns?: ICampaignData[];
}
