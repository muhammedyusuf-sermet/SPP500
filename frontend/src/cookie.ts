import { API_URL } from "./config";
import Store from 'electron-store';

export class CookieManager {
	private constructor(){
	}

	private static store = new Store<string>();

	public static UserToken(cookieToCheck:string):string {
		// undefined if user is not logged in.
		var cookieVal = CookieManager.store.get(cookieToCheck);

		if (cookieVal === undefined)
			return cookieVal

		var request = require("request");
		var options = { method: 'GET',
			url: API_URL + '/verify',
			timeout: 2000,
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json',
				'Authorization': cookieVal
			},
			body: {},
			json: true
		};

		request(options, function (error: string, response: string, body: {status: number, message: string}) {
			if (error) {
				return undefined
			}
			else {
				return (body.status === 201) ? "Valid cookie" : undefined // Success
			}
		});
		return CookieManager.store.get(cookieToCheck);
	}

	public static SetStringCookie(name:string, data:string) {
		CookieManager.store.set(name, data);
	}

	public static RemoveCookie(name:string) {
		CookieManager.store.delete(name);
	}
}
