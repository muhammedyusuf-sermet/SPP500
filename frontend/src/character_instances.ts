import { CharacterRace, CharacterClass, ICharacterData } from "./character";


let characterExample: ICharacterData;
characterExample = {
	Id: 1,
	Name: "Character",
	Level: 3,
	Race: CharacterRace.Gnome,
	Class: CharacterClass.Paladin,
	MaxHealth: 23,
	ArmorClass: 12,
	Notes: 'aggressive',

	Campaigns: []
};

export const CharacterInstances: ICharacterData[] = new Array();

// Generate a dummy array of characters with different IDs
for (var i = 0; i < 12; i++) {
	let character = {} as ICharacterData;
	Object.assign(character, characterExample);
	character.Id = i + 1;
	CharacterInstances.push(character);
}