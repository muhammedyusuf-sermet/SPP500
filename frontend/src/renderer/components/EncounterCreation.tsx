import * as React from 'react';
import * as request from 'request';
import 'bulma/css/bulma.css';
import { Input, Label, Button, Checkbox, Control, Field, TextArea, Modal, ModalContent, ModalBackground, Box} from 'bloomer';
import { Redirect } from "react-router-dom"
import { API_URL } from "../../config"
import { CookieManager } from "../../cookie";

// Todo: Actually get the monsters from the db once the backend is ready
// Dummy array of monsters
/*const monsters = [	{"Id": "1", "Name": "Monster 1"},
					{"Id": "2", "Name": "Monster 2"},
					{"Id": "3", "Name": "Monster 3"}
				];
				*/
const monster = [];

interface IEncounterMonsterInformation {
	"Id": string,
	"Name": string,
}

interface IEncounterResponse {
	status: number,
	messages: string[]
}

interface IMonsterGetResponse {
	status: number,
	messages: string[],
	content: IEncounterMonsterInformation[],
	total: number,
}

export interface IEncounterCreationState {
	redirectToHome: boolean,
	checkedMonsters: Map<string, boolean>,
	page: number,
	pageSize: number,
	totalMonsters: number,
	monstersInCurrentPage: IEncounterMonsterInformation[],
	fullInformationOfCheckedMonsters: IEncounterMonsterInformation[],
	encounter: {
		name: string,
		description: string,
	},
	modal: {
		open: boolean,
		message: string
	}
}

export class EncounterCreation extends React.Component<any, IEncounterCreationState> {
	constructor(props: any) {
		super(props);
		this.state = {
			redirectToHome: false,
			checkedMonsters: new Map(),
			page: 0,
			pageSize: 12,
			totalMonsters: 0,
			monstersInCurrentPage: [] as IEncounterMonsterInformation[],
			fullInformationOfCheckedMonsters: [] as IEncounterMonsterInformation[],
			encounter: {
				name: "",
				description: "",
			},
			modal: {
				open: false,
				message: ""
			}
		}
		this.getPaginatedMonsters(this.state.page);
	}

	previousPage() {
		if(this.state.page > 0){
			let newPage = this.state.page-1;
			this.setState({ page: newPage});
			this.getPaginatedMonsters(newPage);
		}
	}

	nextPage() {
		// Starts from 0
		let totalPages = Math.ceil(this.state.totalMonsters / this.state.pageSize)-1;
		if(this.state.page < totalPages){
			let newPage = this.state.page+1;
			this.setState({ page: newPage});
			this.getPaginatedMonsters(newPage);
		}
	}

	getPaginatedMonsters(page: number) {
		var options = { method: 'GET',
			url: API_URL + '/monster/get/' + page + '/' + this.state.pageSize,
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json' ,
				'Authorization': CookieManager.UserToken('session_token')
			},
			json: true
		};

		request(options, (error:string, responce: any, body: IMonsterGetResponse) => {
			console.log(body);
			if (!error && body.status === 201) { // success
				this.setState({
						monstersInCurrentPage: body.content,
						totalMonsters: body.total,
				});
			} else {
				// There was an error retrieving the monsters. Just return empty array.
				// No need to print a modal.
				this.setState({
						monstersInCurrentPage: [] as IEncounterMonsterInformation[],
						totalMonsters: 0,
				});
			}
		})
	}

	handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const encounter = this.state.encounter
		this.setState({
			encounter: { ...encounter, name: event.target.value }
		})
	}

	handleDescriptionChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const encounter = this.state.encounter
		this.setState({
			encounter: { ...encounter, description: event.target.value }
		})
	}

	handleMonsterCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		event.persist();
		//const id = event.target.attributes['data-id'].value;
		const id = event.target.name;
		const isChecked = event.target.checked;

		this.setState(prevState => ({ checkedMonsters: prevState.checkedMonsters.set(id, isChecked)}));
		console.log(this.state.checkedMonsters);
	}

	openModal = (messageText: string) => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: true, message: messageText }});
	};

	closeModal = () => {
		const modal = this.state.modal
		this.setState({ modal: {...modal, open: false }});
	};

	createEncounter = (event: React.FormEvent) => {
		event.preventDefault();
		this.saveEncounter();
	}

	saveEncounter = (callback?: (message: string) => void) => {
		// Gather the selected monsters for this encounter
		console.log(this.state.checkedMonsters);
		const checkedMonsterIds = [ ...this.state.checkedMonsters.keys() ];
		const keptMonsterIds = [];
		for (var i = 0; i < checkedMonsterIds.length; i++) {
			if (this.state.checkedMonsters.get(checkedMonsterIds[i].toString())) {
				keptMonsterIds.push({"Id": checkedMonsterIds[i]});
			}
		}

		console.log(keptMonsterIds);

		const request = require("request");
		const options = { method: 'POST',
			url: API_URL + '/encounter/create',
			timeout: 2000,
			headers:
			{
				'Cache-Control': 'no-cache',
				'Content-Type': 'application/json',
				'Authorization': CookieManager.UserToken('session_token')
			},
			body:
			{
				Name: this.state.encounter.name,
				Description: this.state.encounter.description,
				Monsters: keptMonsterIds
			},
			json: true
		};
		request(options, (error: string, response: string, body: IEncounterResponse) => {
			if (error) {
				this.openModal("There has been a server error when saving the encounter. Please try again later.");
			} else {
				let { status, messages } = body
				if (status == 201){
					this.openModal("Hooray, the encounter is successfully saved!");
					this.setState({ redirectToHome: true});
				}
				else{
					let message = "There has been an error saving the encounter. Please check your input and try again."
					if (messages) {
						message = messages.join(' ');
					}
					this.openModal(message);
				}
			}
		});
	}

	cancel = (event: React.FormEvent) => {
		event.preventDefault();
		this.setState({ redirectToHome: true});
	}

	render() {
		return (
			(this.state.redirectToHome && !this.state.modal.open) ? <Redirect to="/"/> :
			<div>
				<form id="createEncounterForm" onSubmit={this.createEncounter}>
					<Field>
						<Label>Encounter Name</Label>
						<Control>
							<Input
									id="encounter_name"
									type="text"
									placeholder='Please enter the name of the encounter.'
									onChange={this.handleNameChange} />
						</Control>
					</Field>

					<Field>
					    <Label>Description</Label>
					    <Control>
					        <TextArea
					        			id="encounter_description"
					        			placeholder={'Please write the encounter description here.'}
					        			onChange={this.handleDescriptionChange} />
					    </Control>
					</Field>

					<Field>
						<Control>
							{this.state.monstersInCurrentPage.map(monster => (
								<Checkbox 	key={monster.Id}
											name={monster.Id}
											checked={this.state.checkedMonsters.get(monster.Id)}
											onChange={this.handleMonsterCheckboxChange}>
									{monster.Name}
								</Checkbox>
							))}
						</Control>
					</Field>
					
					<h3>Page No: {this.state.page+1}</h3>
					<a onClick={() => this.previousPage()} id="previousPageButton" className="previous">&laquo; Previous</a>
					<a onClick={() => this.nextPage()} id="nextPageButton" className="next">Next &raquo;</a>
					

					<Field isGrouped>
					    <Control>
					        <Button isColor='primary' type="submit">Submit</Button>
					    </Control>
					    <Control>
					        <Button id="cancel" isLink onClick={this.cancel}>Cancel</Button>
					    </Control>
					</Field>
				</form>
				<Modal id='encounterCreationModal' isActive={this.state.modal.open}>
					<ModalBackground id='modalBackground' onClick={()=>{
						this.closeModal();
					}}/>
					<ModalContent>
						<Box>
							<span id="ModalMessage">{this.state.modal.message}</span>
						</Box>
					</ModalContent>
				</Modal>
			</div>
		);
	}
}