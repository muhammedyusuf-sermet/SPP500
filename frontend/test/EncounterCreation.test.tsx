import * as React from "react"
import { shallow, ShallowWrapper } from 'enzyme';
import { EncounterCreation, IEncounterCreationState } from "../src/renderer/components/EncounterCreation";
//import { Checkbox } from 'bloomer';

import * as nock from 'nock';
import {API_URL} from '../src/config'

jest.mock('../src/cookie');

describe('Encounter Creation', () => {

	let encounterCreationInstance: ShallowWrapper<any, IEncounterCreationState, EncounterCreation>;

	describe('Happy Path', () => {

		console.log(API_URL);
		beforeEach(() => {
			encounterCreationInstance = shallow (<EncounterCreation/>);
		})

		it('renders without crashing', () => {
			expect(encounterCreationInstance).toBeDefined();
		});

		it('renders correctly when the page is loaded', () => {
			expect(encounterCreationInstance).toMatchSnapshot();
		});

		it('should allow text boxes respond to change event and update the state', () => {
			var nameBox = encounterCreationInstance.find('#encounter_name');
			var descriptionBox = encounterCreationInstance.find('#encounter_description');

			nameBox.simulate('change', {target: {name: 'name', value: 'test_name'}});
			descriptionBox.simulate('change', {target: {name: 'description', value: 'test_description'}});

			expect(encounterCreationInstance.state('encounter')).toEqual({
				name: 'test_name',
				description: 'test_description'
			});
		});

		it('should redirect to main page when Cancel button is clicked', () => {
			var cancelButton = encounterCreationInstance.find('#cancel');

			cancelButton.simulate('click', {preventDefault: () => {}});

			expect(encounterCreationInstance.state('redirectToHome')).toEqual(true);
		});

		describe('interacts with the DB data accordingly', () => {

			beforeEach(() => {
				nock.disableNetConnect();
			nock(API_URL)
				.get('/monster/get/0/12')
				.delayBody(2000)
				.reply(201,
					{ "status": 201,
					"messages": [],
					"content": [ {"Id": "1", "Name": "Monster 1"},
						{"Id": "2", "Name": "Monster 2"},
						{"Id": "3", "Name": "Monster 3"},
						{"Id": "4", "Name": "Monster 4"},
						{"Id": "5", "Name": "Monster 5"},
						{"Id": "6", "Name": "Monster 6"},
						{"Id": "7", "Name": "Monster 7"},
						{"Id": "8", "Name": "Monster 8"},
						{"Id": "9", "Name": "Monster 9"},
						{"Id": "10", "Name": "Monster 10"},
						{"Id": "11", "Name": "Monster 11"},
						{"Id": "12", "Name": "Monster 12"},
						{"Id": "13", "Name": "Monster 13"},
						{"Id": "14", "Name": "Monster 14"},
						{"Id": "15", "Name": "Monster 15"},
					],
						"total": "15", 
					},
				).log(console.log);
			encounterCreationInstance = shallow (<EncounterCreation/>);
			});

			it('should allow monster checkboxes respond to change event and update the state', async () => {
				
				//const response = await encounterCreationInstance.state(0, 12);

				//expect(getApiCall(0, 12)).toEqual([1]);
				/*
				var checkboxMonster = encounterCreationInstance.find(Checkbox);

				console.log("hello!");
				// Check a monster, doesn't matter which one
				console.log( encounterCreationInstance.state('monstersInCurrentPage'));

				
				checkboxMonster.at(0).simulate('change',{ target: { checked: true }, persist: () => {} });

				expect(encounterCreationInstance.state('checkedMonsters')).not.toEqual(new Map()); 
				*/
			});
		})

		describe('makes a server request to save the encounter', () => {

			beforeEach(() => {
				nock.disableNetConnect();
				nock(API_URL)
				.post('/encounter/create', {
						Name: "test_name",
						Description: "test_description",
						Monsters   : ["test_monster1"]
					})
				.reply(201,
					{ status: 201, messages: 'success' },
				);
			});

			it('successfully saves the encounter', () => {
				// Since monsters aren't required, this test does not select any monster for the encounter
				var nameBox = encounterCreationInstance.find('#encounter_name');
				var descriptionBox = encounterCreationInstance.find('#encounter_description');
				var createEncounterForm = encounterCreationInstance.find('form');

				nameBox.simulate('change', {target: {name: 'name', value: 'test_name'}});
				descriptionBox.simulate('change', {target: {name: 'description', value: 'test_description'}});

				nock(API_URL)
					.post('/encounter/create', {
							Name: "test_name",
							Description: "test_description",
							Monsters   : ["test_monster1"]
						})
					.reply(201, {
						body: [{ status: 201, messages: 'success' }],
					});
				createEncounterForm.simulate('submit', {preventDefault: () => {}});
			});
		});

		describe('should show and hide modal', () => {
			it('show modal', () => {
				encounterCreationInstance.instance().closeModal();
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(false);
				encounterCreationInstance.instance().openModal('TestMessage');
				expect(encounterCreationInstance.find('#ModalMessage').text()).toEqual('TestMessage');
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(true);
			})

			it('close modal', () => {
				encounterCreationInstance.instance().openModal('TestMessage');
				expect(encounterCreationInstance.find('#ModalMessage').text()).toEqual('TestMessage');
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(true);
				encounterCreationInstance.instance().closeModal();
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(false);
			})

			it('close modal by click', () => {
				encounterCreationInstance.instance().openModal('TestMessage');
				expect(encounterCreationInstance.find('#ModalMessage').text().length).toBeGreaterThan(0);
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(true);
				let background = encounterCreationInstance.find('#modalBackground')
				background.simulate('click');
				expect(encounterCreationInstance.find('#encounterCreationModal').prop('isActive')).toEqual(false);
			})
		});

	});

});