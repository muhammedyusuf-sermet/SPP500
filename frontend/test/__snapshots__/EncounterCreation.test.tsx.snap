// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encounter Creation Happy Path renders correctly when the page is loaded 1`] = `
<div>
  <form
    id="createEncounterForm"
    onSubmit={[Function]}
  >
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Label)>
        Encounter Name
      </withHelpersModifiers(Label)>
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(Input)
          id="encounter_name"
          onChange={[Function]}
          placeholder="Please enter the name of the encounter."
          type="text"
        />
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Label)>
        Description
      </withHelpersModifiers(Label)>
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(TextArea)
          id="encounter_description"
          onChange={[Function]}
          placeholder="Please write the encounter description here."
        />
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
    <withHelpersModifiers(Field)>
      <withHelpersModifiers(Control) />
    </withHelpersModifiers(Field)>
    <h3>
      Page No: 
      1
    </h3>
    <a
      className="previous"
      id="previousPageButton"
      onClick={[Function]}
    >
      « Previous
    </a>
    <a
      className="next"
      id="nextPageButton"
      onClick={[Function]}
    >
      Next »
    </a>
    <withHelpersModifiers(Field)
      isGrouped={true}
    >
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(Button)
          isColor="primary"
          type="submit"
        >
          Submit
        </withHelpersModifiers(Button)>
      </withHelpersModifiers(Control)>
      <withHelpersModifiers(Control)>
        <withHelpersModifiers(Button)
          id="cancel"
          isLink={true}
          onClick={[Function]}
        >
          Cancel
        </withHelpersModifiers(Button)>
      </withHelpersModifiers(Control)>
    </withHelpersModifiers(Field)>
  </form>
  <withHelpersModifiers(Modal)
    id="encounterCreationModal"
    isActive={false}
  >
    <withHelpersModifiers(ModalBackground)
      id="modalBackground"
      onClick={[Function]}
    />
    <withHelpersModifiers(ModalContent)>
      <withHelpersModifiers(Box)>
        <span
          id="ModalMessage"
        />
      </withHelpersModifiers(Box)>
    </withHelpersModifiers(ModalContent)>
  </withHelpersModifiers(Modal)>
</div>
`;
